### PR TITLE
Some Cubic Clean Up

### PR DESCRIPTION
**Pull request by nibanks**
Changes time values (except `kCubic`) to microseconds instead of milliseconds, since they're already 64-bit (which can represent something like half a million years worth of microseconds). Also tries to update some of the calculations a bit to better handle limits/overflows at different stages, but still not exactly happy with it. I want to talk with @maolson-msft some on this.

### CODE REVIEW

File: src/core/cubic.c [hunk 1]
```diff
@@ -460,9 +457,14 @@ CubicCongestionControlOnDataAcknowledged(
...
-            (int64_t)Cubic->KCubic +
-            (int64_t)US_TO_MS(AckEvent->SmoothedRtt);
+            US_TO_MS(
+                (int64_t)TimeInCongAvoidUs -
+                (int64_t)MS_TO_US(Cubic->KCubic) +
+                (int64_t)AckEvent->SmoothedRtt
+            );
+        if (DeltaT > 2500000) {
+            DeltaT = 2500000;
+        }
```
[the remark this answers is not in the document]
Comment (maolson-msft) [reply src/core/cubic.c:468]: "not in the hot path ever"- can you elaborate? This is the ACK handler, which I would certainly call a hot path of this module. Also, exact reproduction of a cubic function isn't critical here; we mainly just need an S-shaped curve, which this accomplishes.

Also, what is the benefit of changing things from msec to usec? K is still in msec so it's not like we are consistently using one time unit. In fact, things were more consistent (everything in cubic was in millisec AFAICT) before this change.

File: src/core/cubic.c [hunk 2]
```diff
@@ -460,9 +458,14 @@ CubicCongestionControlOnDataAcknowledged(
...
         int64_t DeltaT =
-            (int64_t)TimeInCongAvoid -
-            (int64_t)Cubic->KCubic +
-            (int64_t)US_TO_MS(AckEvent->SmoothedRtt);
+            US_TO_MS(
+                (int64_t)TimeInCongAvoidUs -
+                (int64_t)MS_TO_US(Cubic->KCubic) +
+                (int64_t)AckEvent->SmoothedRtt
+            );
+        if (DeltaT > 2500000) {
```
Comment (maolson-msft) [inline src/core/cubic.c:466]: It seems like you've refined the overflow check, but where did the number 2.5M come from?

Comment (nibanks) [reply src/core/cubic.c:466]: I grabbed this number from the comment above (which I believe you wrote?), and did some math to validate that the number did seem to work. I'm still not particularly happy with it though. Also, with the limit here, I'm not sure we need the one below, checking for a negative `int64_t` any more. Going even further, the code at the bottom here limits cwnd to 2 * max_bytes_in_flight, which I think we might be able to be smarter about and elide more calculations if threshold is over a limit.

File: src/core/cubic.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (maolson-msft) [reply src/core/cubic.c:468]: My feelings here are the same as in our other comment thread, which is that if a change is going to increase the difference between the QUIC and TCP cubic code, there ought to be a serious need for it. And the time scale of the cubic function (which tends to have K values around 5-10sec in my experience) is such that microsecond granularity isn't needed, AFAICT. If you think of a specific reason why microsecond granularity will improve things, of course, that's a different matter.

[the remark this answers is not in the document]
Comment (maolson-msft) [reply src/core/cubic.c:468]: (K gives roughly the width of the flat region of the cubic curve, i.e. the time period during which CongestionWindow hovers around WindowMax, where WindowMax is the current estimate of the ideal window, before ramping up to probe for available bandwidth. This is the case unless we are in the aimd-friendly region.)

**Comment by maolson-msft** [comment]
I'm still skeptical that these are improvements, but I don't think the new differences from the TCP code are too much since, say, the core cubic function calculation is still in the same units. Let's merge this as-is. @nibanks

**Review by maolson-msft — APPROVED** [review]

**Pull request merged**